docs/codeql/ql-language-reference/ql-language-specification.rst

```diff
@@ -122,14 +122,23 @@ Global environments
 
 The global module environment has a single entry ``QlBuiltins``.
 
-The global type environment has entries for the primitive types ``int``, ``float``, ``string``, ``boolean``, and ``date``, as well as any types defined in the database schema.
+The global type environment has entries for the primitive types ``int``, ``float``, ``string``, ``boolean``, and ``date``.
 
-The global predicate environment includes all the built-in classless predicates, as well as any extensional predicates declared in the database schema.
+The global predicate environment includes all the built-in classless predicates.
 
 The three global signature environments are empty.
 
 The program is invalid if any of these environments is not definite.
 
+Database schema environments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The database schema type environment has entries for types declared in the database schema.
+
+The database schema predicate environment has entries for extensional predicates declared in the database schema.
+
+The program is invalid if any of these environments is not definite.
```

Comment:
I don't actually remember the significance of this line. Why do we state this explicitly, which case are we excluding here?

Reply:
We are excluding cases where the database schema defines multiple types or predicates with the same name.

Reply:
Hm, seems like "definite" is a very vague way of stating that. However, I see that term used several times, so this PR is probably not the place to change it.
I don't think we need it above in the section on global names though - we control them completely via the compiler.

Reply:
The name resolution section does define the meaning:
> A definite environment has at most one entry for each key. Resolution is unique in a definite environment.

Leaving it out means that the semantics is still defined in cases where the database schema has multiple entries with the same name, even if that will never occur in practice. That seems unnecessary and since it will never occur in practice, surely it is easier to just specify that the semantics is not defined if it occurs.

Reply:
What I meant is that now that we split the built-in names from the dbschema names, this restriction is specifically on the dbschema names. We do not need it for the global names, which at this point is just the built-in names.

Reply:
I see. Agreed, I will change that.

Reply:
Thanks for pointing out the definition, that definitely makes it the right phrasing to use then!